src/s3path.jl

```diff
@@ -0,0 +1,235 @@
+struct S3Path <: AbstractPath
+    segments::Tuple{Vararg{String}}
+    root::String
+    drive::String
+    dir::Bool
+    config::AWSConfig
+end
+
+function S3Path()
+    config = aws_config()
+    account_id = aws_account_number(config)
+    region = config[:region]
+
+    return S3Path(
+        (),
+        "/",
+        "s3://$account_id-$region",
+        true,
+        config,
+    )
+end
+
+function S3Path(str::AbstractString; config::AWSConfig=aws_config())
+    str = String(str)
+    root = ""
+    path = ()
+    dir = true
+
+    @assert startswith(str, "s3://")
+    tokenized = split(str, "/")
+    bucket = strip(tokenized[3], '/')
+    drive = "s3://$bucket"
+
+    if length(tokenized) > 3
+        root = "/"
+        # If the last tokenized element is an empty string then we've parsed a directory
+        dir = isempty(last(tokenized))
+        path = tuple(filter!(!isempty, tokenized[4:end])...)
+    end
+
+    return S3Path(path, root, drive, dir, config)
+end
+
+Base.print(io::IO, fp::S3Path) = print(io, fp.anchor * fp.key)
+
+function Base.:(==)(a::S3Path, b::S3Path)
+    return a.segments == b.segments &&
+        a.root == b.root &&
+        a.drive == b.drive
+end
+
+function Base.getproperty(fp::S3Path, attr::Symbol)
+    if isdefined(fp, attr)
+        return getfield(fp, attr)
+    elseif attr === :anchor
+        return fp.drive * fp.root
+    elseif attr === :separator
+        return "/"
+    elseif attr === :bucket
+        return split(fp.drive, "//")[2]
+    elseif attr === :key
+        return fp.dir ? join(fp.segments, '/') * "/" : join(fp.segments, '/')
+    else
+        # Call getfield even though we know it'll error
+        # so the message is consistent.
+        return getfield(fp, attr)
+    end
+end
+
+# We need to special case join and parents so that we propagate
+# directories correctly
+function Base.join(prefix::S3Path, pieces::AbstractString...)
+    segments = String[prefix.segments...]
+    dir = endswith(last(pieces), "/")
+
+    for p in pieces
+        push!(segments, filter!(!isempty, split(p, "/"))...)
+    end
+
+    return S3Path(
+        tuple(segments...),
+        prefix.root,
+        prefix.drive,
+        dir,
+        prefix.config,
+    )
+end
+
+function parents(fp::S3Path)
+    if hasparent(fp)
+        return map(1:length(fp.segments)-1) do i
+            S3Path(fp.segments[1:i], fp.root, fp.drive, true, fp.config)
+        end
+    else
+        error("$fp has no parents")
```

Comment:
Why doesn't this return itself.
after all `..` in the unix root directory leavies you in in the unix root directory

Reply:
NOTE: This is also an issue with FilePathsBase. https://github.com/rofinn/FilePathsBase.jl/issues/19